### PR TITLE
[feature] File notifications

### PR DESCRIPTION
**Pull request by Ghalko**
This should be considered before wiki_notify #2378 

## Proposal
Individual notify settings on files. Allows granular and project-wide file notification subscription.

## Repercussions
- Adds an API route for retrieving file subscriptions
 - Add to api v2?
- Adds schema validation rather than one-to-one character checking for formatting subscription data for treebeard. Using new python module 'schema' to generate shemas that the data is compared to.
- Adds BaseEvent which is an abstract class for any event.
 - Using the abstract class for a FileEvent and subclasses that take care of parsing the data from file actions.

### CODE REVIEW

**Comment by chennan47** [comment]
@Ghalko This pr needs an update and also needs fix the failing tests.

**Comment by Ghalko** [comment]
The failed tests are from the flake8 issues, should be fixed when I update the merge issues.

**Comment by GaryKriebel** [comment]
When uploading a file, the email notification sent has a slash preceding the file name.
screen shot 2015-06-15 at 2 03 25 pm

This slash should be removed as this is not the same behaviour as elsewhere in the site.
screen shot 2015-06-15 at 2 04 00 pm

**Comment by GaryKriebel** [comment]
When B. Gustafsson multiple files, there was one email notification sent for each file uploaded.  

screen shot 2015-06-15 at 2 19 34 pm

screen shot 2015-06-15 at 2 21 22 pm

screen shot 2015-06-15 at 2 21 48 pm

screen shot 2015-06-15 at 2 22 09 pm


Can they not be grouped together and have one row for each file in the table of one email?  Perhaps like:
screen shot 2015-06-15 at 2 24 26 pm

**Comment by GaryKriebel** [comment]
When creating a folder, no need for the beginning & ending slash.
screen shot 2015-06-15 at 2 40 08 pm

**Comment by GaryKriebel** [comment]
Moving files looks good:
screen shot 2015-06-15 at 2 42 17 pm

as does deleting:
screen shot 2015-06-15 at 2 44 24 pm

**Comment by lyndsysimon** [comment]
@GaryKriebel - Is this PR ready for first-pass code review? It looks like you've performed a functional review, but I'm unsure if all the issues you've raised have been addressed.

**Comment by GaryKriebel** [comment]
I'm not sure if @Ghalko updated the code... the consolidation of emails may need some discussion.  The removal of slashes may be trivial? A first pass on what is there may not be bad.

**Comment by Ghalko** [comment]
@lyndsysimon @GaryKriebel I am fixing things, it will be ready in moments.

**Comment by lyndsysimon** [comment]
I'd prefer a single PR unless you have a strong preference otherwise.

**Comment by lyndsysimon** [comment]
Just let me know when it's ready :)

**Comment by Ghalko** [comment]
The consolidation has almost nothing to do with file notification specifically. Essentially I would have to figure out the best approach to consolidation.
- Use the current digest code by bringing in the digest script and have it set at two set intervals.
- Make a new digest type that would send an Email five minutes after the first change, then reset the timer (essentially making a new digest class of Email)
 - sent five minutes after the last change.

Both options would be feasible, and could eliminate the normal Email type notification. That being said, if a file or project is being heavily used and someone signed up to spam themselves with every single change, the last option might cause an Email to never be sent if people all over the world are working constantly on the files within a project. Personally I think this is outside the scope of file notifications. This should be considered for all notifications.

File: tests/test_addons.py [hunk 1]
```diff
@@ -217,7 +217,8 @@ def build_payload(self, metadata, **kwargs):
             'signature': signature,
         }
 
-    def test_add_log(self):
+    @mock.patch('website.addons.base.notifications.file_notify')
```
Comment (lyndsysimon) [inline tests/test_addons.py]: If we're mocking `file_notify`, shouldn't we also assert that it was called?

I get that this is testing that a log was added, but I think I'd rather the test fail when the log functionality is broken than later have the mock be unnecessary and the test pass.

**Comment by lyndsysimon** [comment]
After speaking in depth, I've asked that @Ghalko refactor `/website/addons/base/nofications.py` from a series of nested function calls (wrapped by `file_notify`), to a class hierarchy.

I believe this will take approximately the same amount of time as documenting the behavior of the code as it is, and will help make the flow more clear. It will also allow us to reuse a great deal of the logic for future notifications, potentially even those outside the context of files.

**Comment by GaryKriebel** [comment]
This should be good for another review... functional review looked good.

**Comment by Ghalko** [comment]
I am straightening out celery issues on #3200. I guess I should close this one because #3200 is based on it, so they have all the same commits.

**Comment by lbanner** [comment]
@Ghalko Clarifying: #3200 now supersedes and includes this PR?

**Comment by Ghalko** [comment]
@lbanner Correct.

**Comment by lyndsysimon** [comment]
Closing this in favor of #3200.

**Pull request closed without merging**